util.c

```diff
@@ -5322,3 +5286,159 @@ int util_is_wlan(char *device)
   return slist_getentry(config.net.wlan.devices, device) ? 1 : 0;
 }
 
+
+/*
+ * Write log message.
+ *
+ * level is a bitmask determining the destination to log to.
+ *
+ * Add a time stamp when the destination has the LOG_TIMESTAMP flag set.
+ */
+void util_log(unsigned level, char *format, ...)
+{
+  va_list args;
+  char *buf, *caller = NULL;
+  int buf_len = 0;
+  log_file_t *lf;
+
+  time_t t = time(NULL);
+  struct tm *gm = gmtime(&t);
+
+  va_start(args, format);
+  if(vasprintf(&buf, format, args) == -1) buf = NULL;
+  if(buf) buf_len = strlen(buf);
+  va_end(args);
+
+  for(lf = config.log.dest; lf < config.log.dest + sizeof config.log.dest / sizeof *config.log.dest; lf++) {
+    if((level & lf->level)) {
+      if(!lf->f && lf->name) {
+        lf->f = fopen(lf->name, "a");
+      }
+      if(lf->f) {
+        if((lf->level & LOG_TIMESTAMP)) {
+          if(gm) {
+            fprintf(lf->f, "%02d:%02d:%02d <%u>", gm->tm_hour, gm->tm_min, gm->tm_sec, level);
+            if((lf->level & LOG_CALLER)) {
+              if(!caller) caller = util_get_caller(1);
+              if(caller) fprintf(lf->f, " %-28s", caller);
+            }
+            fprintf(lf->f, ": ");
+          }
+        }
+        if(buf_len) {
+          fwrite(buf, buf_len, 1, lf->f);
+          // supplement a missing trailing newline
+          if(
+            buf[buf_len - 1] != '\n' &&
+            (lf->level & LOG_TIMESTAMP)
+          ) {
+            fputc('\n', lf->f);
```

Comment:
so it automatic add newline if printed with timestamp...quite confusing for me...why not if there is caller only?